### ISSUE TITLE
mathjax not rendering

### ISSUE THREAD

**Issue by jwhendy** [rstudio/shiny-examples#49]
I posted on SO as well [here](http://stackoverflow.com/questions/40143012/r-shiny-mathjax-not-rendering-math). Here's the question:

---

I'm having odd behavior with `MathJax` and `shiny`. I've modified the `01_hello` example from the `shiny` package to include a `.Rmd` file via `includeMarkdown()`. All I modified was `ui.R` to be as follows:

```
# Show a plot of the generated distribution
mainPanel(
  plotOutput("distPlot"),
  includeMarkdown("mathjax-test.Rmd")
)
```

The contents of my test file (`mathjax-test.Rmd`) are like so:

```
## mathjax test

Document used as is in .Rmd format.

Here's an inline equation with dollar signs: $a + b = c$

Here's an inline equation with slashes/parentheses: \(a + b = c\)

Display format wth double dollar signs:

$$a + b = c$$

Display format with slashes/square brackets:

\[
a + b = c
\]
```

When I knit the `.Rmd` file, I get the expected results:

[![knit-rmd](https://i.stack.imgur.com/Gyrw4.png)](https://i.stack.imgur.com/Gyrw4.png)

When I use `runApp()` I get this:

[![shiny-rmd](https://i.stack.imgur.com/z3EwB.png)](https://i.stack.imgur.com/z3EwB.png)

---

Things I've tried:
- adding `withMathJax()` after `titlePanel` in `ui.R` per this RStudio [page](http://shiny.rstudio.com/gallery/mathjax.html)
- wrapping `includeMarkdown()` inside of `withMathJax()`
- putting `library(shiny)` and `withMathJax()` in a code block at the top of the `.Rmd` file
- adding the following first, right before, and right after `includeMarkdown()` in the `mainPanel` section of `ui.R`:

`tags$script(src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML")`

I have no idea what's going on! I get this in the `<head>` of the `shiny` app page when I view the source:

```
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script>
```

What am I doing wrong?

**Comment by guiastrennec** [comment]
Not sure if you have solved your issue but try this minimal (app.R) example, it should give you the result that you want.

```r
library(shiny)
shinyApp(ui = fluidPage(
  withMathJax(includeMarkdown('mathjax-test.Rmd'))), 
  server = function(input, output) { })
```

**Comment by jwhendy** [comment]
@guiastrennec Did you try it with my `mathjax-test.Rmd` file above? If so, could you kindly post a screenshot? *Something* might have updated, as it appears both display math variants render properly (`\[ math \]` and `$$ math $$`). Both inline versions (`$ math $` and `\( math \)`) still fail:

![2016-11-29_212410](https://cloud.githubusercontent.com/assets/379357/20738650/506fc480-b67a-11e6-8494-146cfaf09d29.png)

I'm curious if you get the same result. If not (and yours renders properly), I'll have to dig deeper I guess; could this be system/setup dependent?

**Comment by guiastrennec** [comment]
Yes the `mathjax-test.Rmd` was copy-pasted from your example above (note that I did not include a YAML header). I assume that this issue has been corrected in a recent package update.

Here is a screenshot of what I get when I click the `Run App` button in Rstudio:
<img width="417" alt="mathjax_shiny" src="https://cloud.githubusercontent.com/assets/7304883/20746335/a5ddb54c-b6e5-11e6-8339-efffd68e7281.png">

And here is my `sessionInfo()`:
```r
R version 3.3.2 (2016-10-31)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X El Capitan 10.11.6

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_0.14.2

loaded via a namespace (and not attached):
 [1] R6_2.2.0        markdown_0.7.7  htmltools_0.3.5 tools_3.3.2     Rcpp_0.12.8     jsonlite_1.1   
 [7] digest_0.6.10   xtable_1.8-2    httpuv_1.3.3    mime_0.5       
```

**Comment by jwhendy** [comment]
Output of `sessionInfo()` as I would have had above:
```
R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Arch Linux

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=C              
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_0.13.2

loaded via a namespace (and not attached):
[1] R6_2.1.2        rsconnect_0.4.3 htmltools_0.3.5 tools_3.3.2     Rcpp_0.12.7     digest_0.6.10   xtable_1.8-2   
[8] httpuv_1.3.3    mime_0.5  
```

After updating just now:
```
R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Arch Linux

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=C              
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_0.14.2

loaded via a namespace (and not attached):
 [1] R6_2.2.0        rsconnect_0.6   markdown_0.7.7  htmltools_0.3.5 tools_3.3.2     Rcpp_0.12.8     jsonlite_1.1   
 [8] digest_0.6.10   xtable_1.8-2    httpuv_1.3.3    mime_0.5
```

After updating I closed RStudio, reopened, loaded `shiny` and get the same result. Some observations in looking at the `shiny` vs. direct `knitr` page sources:

- scripts are embedded with `knitr` and linked in `shiny`. That said, `shiny` can still render *some* math, so I don't think that should be an issue.

- the math itself is different between the two. The `shiny` source looks like this:
```
<p>Here&#39;s an inline equation with dollar signs: $a + b = c$</p>
<p>Here&#39;s an inline equation with slashes/parentheses: (a + b = c)</p>
```

The `knitr` output looks like this (note that somewhere along the way it converted `$` to `\(` style):
```
<p>Here’s an inline equation with dollar signs: <span class="math inline">\(a + b = c\)</span></p>
<p>Here’s an inline equation with slashes/parentheses: <span class="math inline">\(a + b = c\)</span></p>
```

- I caught an inconsistency between the example and my continued fiddling after posting it. It appears my local version featured `\\[ math \\]` for my second display math format, which is *not* rendering properly after changing it to just `\[ math \]`. Sorry about that.

I thought I'd just create a [repo](https://github.com/jwhendy/shiny-mathjax-test) for this test. It has the full `mathjax-test.Rmd` file and screenshots of what I get via `shiny` vs. `knitr`. I also saved out the "view source" results from the `shiny` app in case you'd want to compare yours and mine.

**Comment by guiastrennec** [comment]
I tried your example from the repo and it's working properly with both `knitr` and `runApp()`. Here is the screenshot:
<img width="589" alt="mathjax-test-2" src="https://cloud.githubusercontent.com/assets/7304883/20773625/00bfcd0c-b753-11e6-8f74-36acd8cbb1de.png">

I am sorry but I don't know what could be the problem, maybe something to do with linux or your locale?! Could you test it on a different machine?

**Comment by jwhendy** [comment]
Ha. Yes, I was starting to suspect locale earlier today as well; that's all that's sticking out for differences.

Do you get output from that code block at the end? For me, `knitr` outputs both the block and the formatted math. Is that the same for you? Doesn't really matter as it's not my main issue, but just curious if we replicate on that point as well.

**Comment by guiastrennec** [comment]
Yes `knitr` runs R in the chunk and include the output in the html, however in shiny the `Rmd` file is read as `markdown` and hence the code block are not run. I am actually not sure whether we are supposed to use .Rmd in the `includeMarkdown()` function, however, I also tried to use this YAML header in the .Rmd, knit it and use the .md file into the app but it does not seem to be compatible with equations.

```
---
output: md_document
---
```

**Comment by jwhendy** [comment]
Trying to check on the locale possibility. I'm a little confused at how this is going down.

```
$ locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE=C
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

I checked on the [docs for `sessionInfo()`](https://stat.ethz.ch/R-manual/R-devel/library/utils/html/sessionInfo.html) and it says that contains the output of `Sys.getlocale()`, so here's that:
```
> Sys.getlocale()
[1] "LC_CTYPE=en_US.UTF-8;LC_NUMERIC=C;LC_TIME=en_US.UTF-8;
LC_COLLATE=C;LC_MONETARY=en_US.UTF-8;LC_MESSAGES=en_US.UTF-8;
LC_PAPER=en_US.UTF-;LC_NAME=C;LC_ADDRESS=C;LC_TELEPHONE=C;
LC_MEASUREMENT=en_US.UTF-8;LC_IDENTIFICATION=C"
```

As you can see, what `R` is picking up isn't matching my system. I was able to use `Sys.setlocale(category = "blah", locale = "en_US.UTF-8")` to fix `LC_NUMERIC`, but attempts to change the others set to `C` failed with the message:
```
> Sys.setlocale(category = "LC_ADDRESS", locale = "en_US.UTF-8")
Error in Sys.setlocale(category = "LC_ADDRESS", locale = "en_US.UTF-8") : 
  invalid 'category' argument
```

Same result and no idea where to go from here. Would you want to post your `shiny` page source as an `html` attachment? Maybe I can dig through that to see if there are any differences for clues on what's being generated.

**Comment by guiastrennec** [comment]
I have uploaded the app on shinyapps.io ([here](https://guiastrennec.shinyapps.io/mathjax-markdown/)) and it's rendering properly. To test potential browser incompatibilities, I have tried to view the app in safari, chrome, firefox, IE and even on my phone and in all cases all equations were properly rendered. I have also attached the html from my local runApp() if you wish to investigate further.

[mathjax-shiny.html.zip](https://github.com/rstudio/shiny-examples/files/624111/mathjax-shiny.html.zip)

**Comment by guiastrennec** [comment]
Some additional thoughts, maybe try to update your TeX distribution (I noticed on a machine without LaTeX installed that equations were not rendered!

**Comment by jwhendy** [comment]
Interesting... I have had it on my todo list to go through and update from TexLive 2015 to 2016. I'm also finding that it's working online, recalling I had an AWS shiny-server instance running. Shows up [fine](http://52.36.187.23/shiny-mathjax-test/).

I'll try the Tex route as the `locale` output on my AWS ubuntu server looks similar:

```
$ locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

And from `R` (also changing a bunch to `C` instead of what the system says):
```
> Sys.getlocale()
[1] "LC_CTYPE=en_US.UTF-8;LC_NUMERIC=C;LC_TIME=en_US.UTF-8;
LC_COLLATE=en_US.UTF-8;LC_MONETARY=en_US.UTF-8;
LC_MESSAGES=en_US.UTF-8;LC_PAPER=en_US.UTF-8;
LC_NAME=C;LC_ADDRESS=C;LC_TELEPHONE=C;
LC_MEASUREMENT=en_US.UTF-8;LC_IDENTIFICATION=C"
```

Will try the TexLive update and post back.

**Comment by mine-cetinkaya-rundel** [comment]
I am not able to reproduce this issue, I assume it has been resolved. @jwhendy please let us know if not.

**Comment by jwhendy** [comment]
I'll confirm from the original setup (arch linux) later today when I reboot. At the moment I'm in an Ubuntu installation and it rendered properly in chromium, but was rendered verbatim (i.e. `$a + b = c $`) in the RStudio popup browser window. Is that expected?

Regarding the issue in general, some comments:

- we never confirmed what the problem was. I had issues, others appeared not to have these issues. If we close, it would be awesome to track back to something that was changed between then and now which fixed this
- suggestions were made (`locale`, update `TexLive`) that, in my opinion, depart from the information [provided by RStudio](http://rmarkdown.rstudio.com/html_document_format.html#mathjax_equations). If these issues are linked to the issue, it would be awesome to update RStudio docs with troubleshooting information or things to check if your equations don't render
- my take on the RStudio docs is that this should "just work" since no other deps or setup instructions are specified. Is that the case?

Thanks!

**Comment by jwhendy** [comment]
Tried from arch again just now. Reproducible method: clone this repo, `setwd()` to that dir, `library(shiny)`, and `runApp()`.

**RStudio popup browser**
![2017-10-09_112234](https://user-images.githubusercontent.com/379357/31348197-6fcb50cc-ace4-11e7-9944-915ace59c982.png)

**Chromium 61.0.3163.100**
![2017-10-09_112507](https://user-images.githubusercontent.com/379357/31348220-86a4263e-ace4-11e7-886b-fca2386fb408.png)

```
$ uname -a
Linux hostname 4.13.4-1-ARCH #1 SMP PREEMPT Thu Sep 28 08:39:52 CEST 2017 x86_64 GNU/Linux

$ localectl 
   System Locale: LANG=en_US.UTF-8
       VC Keymap: n/a
      X11 Layout: n/a

> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Arch Linux

Matrix products: default
BLAS: /usr/lib/libblas.so.3.7.1
LAPACK: /usr/lib/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_1.0.3

loaded via a namespace (and not attached):
 [1] compiler_3.4.2  R6_2.2.1        markdown_0.8    htmltools_0.3.6 tools_3.4.2     Rcpp_0.12.11   
 [7] jsonlite_1.4    digest_0.6.12   xtable_1.8-2    httpuv_1.3.3    mime_0.5
```

I'll post the same from my Ubuntu setup shortly. Forgot to do that before I rebooted. The behavior is different.

**Comment by jwhendy** [comment]
From Ubuntu 16.04:

**RStudio popup browser**
![2017-10-10_182911](https://user-images.githubusercontent.com/379357/31415421-03e0b112-ade9-11e7-896f-d88b5f80a60c.png)

**Chromium 61.0.3163.100 (correct)**
![2017-10-10_183035](https://user-images.githubusercontent.com/379357/31415439-20b1592c-ade9-11e7-8798-158cbf28a06e.png)

```
$ uname -a
Linux hostname 4.4.0-93-generic #116-Ubuntu SMP Fri Aug 11 21:17:51 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.3 LTS

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.6.0
LAPACK: /usr/lib/lapack/liblapack.so.3.6.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8   
 [6] LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_1.0.5

loaded via a namespace (and not attached):
 [1] compiler_3.4.2  R6_2.2.0        markdown_0.8    htmltools_0.3.6 tools_3.4.2     Rcpp_0.12.9     jsonlite_1.2    digest_0.6.12  
 [9] xtable_1.8-2    httpuv_1.3.5    mime_0.5
```

There are two things that would be awesome:

- [ ] why doesn't this work properly in the RStudio browser?
- [ ] what's causing this to still not work properly in arch linux

Nothing stands out to me above. The main suspect that's been suggested, `locale`, is identical as far as `sessionInfo()` sees (all `utf-8` except numeric, name, adress, identification, telephone).

**Comment by jcheng5** [comment]
We reproduced the issue two weeks ago while testing the latest release of Shiny on RStudio Server Pro. If you're seeing this on RStudio Desktop, I believe the issue is specific to Linux--we haven't seen it reproduced on Mac or Windows. I looked in the RStudio source code and it turns out it has an embedded copy of the same markdown parser that knitr uses (indirectly via the markdown package). This is causing knitr to unintentionally use RStudio's copy, which is missing a custom latex math extension that was added to the markdown copy. The RStudio IDE team is now aware of the problem, although I can't promise any particular ETA for a fix.

**Comment by aldomann** [comment]
Is this still being worked on?

**Comment by jcheng5** [comment]
It was fixed in RStudio 1.2. If you're still seeing this now, let us know what version of RStudio you're using.

**Comment by aldomann** [comment]
Actually I just realized that `withMathJax("\\( x = 2 \\)")` works just fine on a compiled Rmd, but `withMathJax("$x = 2$")` does work on the Shiny itself. Probably an issue with Knitr rather than `shiny::withMathJax()`.

**Comment by Alik-V** [comment]
Running Rstudio 1.4.1717 and the single `$ formula $` issue is still present. Wrapping formulas in `\\( formula \\)` works as intended

**Comment by jwhendy** [comment]
For what it's worth, I *can't* reproduce from my [git repo example](https://github.com/rstudio/shiny-examples/issues/49#issuecomment-335209470) above any longer. @Alik-V I would propose you create a fully reproducible, standalone example and add far more system details, perhaps in a separate issue linked to this.

**From R terminal, opens in Firefox**
![grab_2022-01-14_105526](https://user-images.githubusercontent.com/379357/149554640-80e3bc04-5c5d-4fe8-aad7-97750a845d7d.png)

**From Rstudio, opens in embedded browser**
![grab_2022-01-14_105606](https://user-images.githubusercontent.com/379357/149554673-71027f0c-34b1-490d-8705-e5fcc60d7a9d.png)

I don't remember what I was trying to show with the block at the bottom anymore, but the four various methods all worked.

```
[1] stats     graphics  grDevices utils     datasets 
[6] methods   base     

other attached packages:
[1] shiny_1.7.1

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7       digest_0.6.28    withr_2.4.2     
 [4] later_1.3.0      mime_0.12        R6_2.5.1        
 [7] jsonlite_1.7.2   lifecycle_1.0.1  xtable_1.8-4    
[10] magrittr_2.0.1   cachem_1.0.6     rlang_0.4.12    
[13] promises_1.2.0.1 jquerylib_0.1.4  bslib_0.3.1     
[16] ellipsis_0.3.2   tools_4.1.2      markdown_1.1    
[19] rsconnect_0.8.24 xfun_0.28        httpuv_1.6.3    
[22] fastmap_1.1.0    compiler_4.1.2   htmltools_0.5.2 
[25] sass_0.4.0      
```

The above was from RStudio 1.4.1717, installed via [arch AUR](https://aur.archlinux.org/packages/rstudio-desktop-bin/). I noticed the pkgver has bumped, so I updated to 2021.09.1.372-1 just now, and it still renders as shown above.

Going to close, but of course others are free to describe their specific system and details in a new issue and link here.
